app/position/providers/androidpositionprovider.cpp

```diff
@@ -0,0 +1,277 @@
+
+}
+
+
+AndroidPositionProvider::AndroidPositionProvider( bool fused, QObject *parent )
+  : AbstractPositionProvider( fused ? QStringLiteral( "android_fused" ) : QStringLiteral( "android_gps" ),
+                              QStringLiteral( "internal" ),
+                              fused ? tr( "Internal (fused)" ) : tr( "Internal (gps)" ), parent )
+  , mFused( fused )
+  , mInstanceId( ++sLastInstanceId )
+{
+  __android_log_print( ANDROID_LOG_INFO, "CPP", "[c++] CONSTRUCT" );
+
+  Q_ASSERT( !sInstances.contains( mInstanceId ) );
+  sInstances[mInstanceId] = this;
+
+  // register the native methods
+
+  JNINativeMethod methods[]
+  {
+    {
+      "jniOnPositionUpdated",
+      "(ILandroid/location/Location;Landroid/location/GnssStatus;)V",
+      reinterpret_cast<void *>( jniOnPositionUpdated )
+    }
+  };
+
+  QJniEnvironment javaenv;
+
+  javaenv.registerNativeMethods( "uk/co/lutraconsulting/MMAndroidPosition", methods, 1 );
+
+  // create the Java object
+
+  __android_log_print( ANDROID_LOG_INFO, "CPP", "[c++] create Java object" );
+
+  jobject context = QNativeInterface::QAndroidApplication::context();
+
+  mAndroidPos = QJniObject::callStaticObjectMethod( "uk/co/lutraconsulting/MMAndroidPosition", "createWithJniCallback",
+                "(Landroid/content/Context;ZI)Luk/co/lutraconsulting/MMAndroidPosition;", context, mFused, mInstanceId );
+
+  // Request permissions if needed
+
+  QLocationPermission perm;
+  perm.setAccuracy( QLocationPermission::Precise );
+  if ( qApp->checkPermission( perm ) != Qt::PermissionStatus::Granted )
+  {
+    // if user previously completely denied location, the permissions request dialog
+    // may not even show up and we get denied response again.
+    __android_log_print( ANDROID_LOG_INFO, "CPP", "[c++] going to request permissions" );
+    qApp->requestPermission( perm, [this]( const QPermission & p )
+    {
+      if ( p.status() == Qt::PermissionStatus::Granted )
+      {
+        __android_log_print( ANDROID_LOG_INFO, "CPP", "[c++] permissions granted!" );
+        this->startUpdates();
+      }
+      else
+      {
+        __android_log_print( ANDROID_LOG_INFO, "CPP", "[c++] permissions denied :-(" );
+        // User may have granted permission just for approximate location,
+        // so it would be good to detect that and warn about that: approximate
+        // location has intentionally only ~2km accuracy - too low for any data collection
+        QLocationPermission permApprox;
+        permApprox.setAccuracy( QLocationPermission::Approximate );
+        if ( qApp->checkPermission( permApprox ) == Qt::PermissionStatus::Granted )
+          this->setState( tr( "Approximate location only!" ), State::NoConnection );
+        else
+          this->setState( tr( "No location permissions" ), State::NoConnection );
+      }
+    } );
+    return;
+  }
+
+  // TODO: this should not be needed?
+  AndroidPositionProvider::startUpdates();
```

Comment:
Todo

Reply:
Todo intentionally left there - the position provider interface probably could get a bit of a cleanup. It seems that in some cases, the startUpdates() is called by the client code, but in other times it is not (app startup vs when changing providers). It would be good to unify it.